#! /bin/sh /usr/share/dpatch/dpatch-run
## 01u1backend.dpatch by  <az@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: backported fix for LP: #1161599 for '400' errors with ubuntu1

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' duplicity-0.6.21~/duplicity/backends/u1backend.py duplicity-0.6.21/duplicity/backends/u1backend.py
--- duplicity-0.6.21~/duplicity/backends/u1backend.py	2013-01-24 00:35:56.000000000 +1000
+++ duplicity-0.6.21/duplicity/backends/u1backend.py	2013-05-31 11:19:02.273511212 +1000
@@ -84,6 +84,10 @@
                           % duplicity.util.exception_traceback())
                 if n == globals.num_retries:
                     log.FatalError("Giving up on request after %d attempts, last exception %s" % (n,e))
+
+                if isinstance(body, file):
+                    body.seek(0) # Go to the beginning of the file for the retry
+
                 time.sleep(30)
                 continue
 
@@ -106,6 +110,9 @@
             elif numcode == 404:
                 ecode = log.ErrorCode.backend_not_found
 
+            if isinstance(body, file):
+                body.seek(0) # Go to the beginning of the file for the retry
+
             if n < globals.num_retries:
                 time.sleep(30)
 
@@ -207,10 +214,12 @@
         remote_full = self.content_base + self.quote(content['content_path'])
         log.Info("uploading file %s to location %s" % (remote_filename, remote_full))
 
+        size = os.path.getsize(source_path.name)
         fh=open(source_path.name,'rb')
 
         content_type = 'application/octet-stream'
-        headers = {"Content-Type": content_type}
+        headers = {"Content-Length": str(size),
+                   "Content-Type": content_type}
         resp, content = self.client.request(remote_full,
                                             method="PUT",
                                             body=fh,
